<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Interactive 3-D association landscape for many phenotypes — landscapefast • iphegwas</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Interactive 3-D association landscape for many phenotypes — landscapefast"><meta property="og:description" content="Interactive 3-D association landscape for many phenotypes"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">iphegwas</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/iphegwas.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/georgeg0/iphegwas" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Interactive 3-D association landscape for many phenotypes</h1>
    
    <div class="hidden name"><code>landscapefast.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Interactive 3-D association landscape for many phenotypes</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">landscapefast</span><span class="op">(</span></span>
<span>  <span class="va">d</span>,</span>
<span>  sliceval <span class="op">=</span> <span class="fl">7</span>,</span>
<span>  chromosome <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  pop <span class="op">=</span> <span class="st">"GBR"</span>,</span>
<span>  R2 <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  D <span class="op">=</span> <span class="fl">0.75</span>,</span>
<span>  calculateLD <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  mutualLD <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  <span class="va">phenos</span>,</span>
<span>  genemap <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  geneview <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  levelsdown <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>d</dt>
<dd><p>DataFrame output from processphegwas</p></dd>


<dt>sliceval</dt>
<dd><p>Integer to indicate value of -log10(p) to do the sectionalcut. Usually value &gt; -log10 6 is considered to be significant</p></dd>


<dt>chromosome</dt>
<dd><p>Integer to indicate the chromosome number thats interested, If not given entire chromosome is given</p></dd>


<dt>pop</dt>
<dd><p>The population to select for calculation the LD (default GB)</p></dd>


<dt>R2</dt>
<dd><p>The value to set to calculate LD</p></dd>


<dt>D</dt>
<dd><p>THe value to set to calcualte LD</p></dd>


<dt>calculateLD</dt>
<dd><p>This shoudld be set to true if the calcualte LD logic needed to be added to the plot</p></dd>


<dt>mutualLD</dt>
<dd><p>Calcualte the mutual LD SNP between the phenotypes</p></dd>


<dt>phenos</dt>
<dd><p>Vector of names of dataframes that need to do PheGWAS on. Arrange the dataframe in the order how the the phenotypes should align in y axis
chromosome view the max peak is selected</p></dd>


<dt>genemap</dt>
<dd><p>if true it map SNP's to gene. This takes time as its using BioMart package to map genes.</p></dd>


<dt>geneview</dt>
<dd><p>This checks for the common genes across the section</p></dd>


<dt>levelsdown</dt>
<dd><p>Used to find the independant signals</p></dd>

</dl></div>
    <div id="author">
    <h2>Author</h2>
    <p>George Gittu</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw">if</span> <span class="op">(</span><span class="cn">FALSE</span><span class="op">)</span> <span class="op">{</span></span></span>
<span class="r-in"><span><span class="co"># here y is output from function fastprocessphegwas</span></span></span>
<span class="r-in"><span><span class="co"># 3D landscape-viz of all the phenos across the base pair positions(threshold &gt; -log10(p) 7.5)</span></span></span>
<span class="r-in"><span><span class="co"># Without iPheGWAS you get output in the order you pass phenos</span></span></span>
<span class="r-in"><span><span class="fu">landscapefast</span><span class="op">(</span><span class="va">y</span>,sliceval <span class="op">=</span> <span class="fl">10</span>,phenos <span class="op">=</span><span class="va">phenos</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 3D landscape visualization after applying iPheGWAS</span></span></span>
<span class="r-in"><span><span class="co"># Passing the order from the iPheGWAS function</span></span></span>
<span class="r-in"><span><span class="fu">landscapefast</span><span class="op">(</span><span class="va">y</span>,sliceval <span class="op">=</span> <span class="fl">10</span>,phenos <span class="op">=</span> <span class="fu"><a href="iphegwas.html">iphegwas</a></span><span class="op">(</span><span class="va">phenos</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">#3D landscape visualization of chromosome number 19 (above a threshold of -log10 (p) 7.5)</span></span></span>
<span class="r-in"><span><span class="fu">landscapefast</span><span class="op">(</span><span class="va">y</span>,sliceval <span class="op">=</span> <span class="fl">7.5</span>,chromosome <span class="op">=</span> <span class="fl">19</span>,phenos <span class="op">=</span><span class="va">phenos</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 3D landscape-viz of chromosome 19, gene view active(threshold &gt; -log10(p) 7.5)</span></span></span>
<span class="r-in"><span><span class="fu">landscape</span><span class="op">(</span><span class="va">y</span>,sliceval <span class="op">=</span> <span class="fl">7.5</span>,chromosome <span class="op">=</span> <span class="fl">19</span>, geneview <span class="op">=</span> <span class="cn">TRUE</span>,phenos <span class="op">=</span><span class="va">phenos</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># 3D landscape-viz with calculate-LD and mutual-LD functionality active</span></span></span>
<span class="r-in"><span><span class="fu">landscapefast</span><span class="op">(</span><span class="va">y</span>, sliceval <span class="op">=</span> <span class="fl">30</span>, chromosome <span class="op">=</span> <span class="fl">19</span>,calculateLD<span class="op">=</span> <span class="cn">TRUE</span>,mutualLD <span class="op">=</span> <span class="cn">TRUE</span>,phenos <span class="op">=</span><span class="va">phenos</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="op">}</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Gittu George.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

