
#' HDL GWAS summary dataset (made available from Lipid consortium data files)
#'
#' The dataset 'hdl' is a data.frame object with the summary results of a GWAS analysis.
#'
#' @docType data
#'
#' @usage data(hdl)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item CHR is the chromosome number
#'   \item BP is the position of the chromosome
#'   \item rsid The SNP ID associated with that position and chromsome
#'   \item HDL is the p_value of the trait
#'   \item gene is the gene associated with rsid
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"hdl"

#' LDL GWAS summary dataset (made available from Lipid consortium data files)
#'
#' The dataset 'ldl' is a data.frame object with the summary results of a GWAS analysis.
#'
#' @docType data
#'
#' @usage data(ldl)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item CHR is the chromosome number
#'   \item BP is the position of the chromosome
#'   \item rsid The SNP ID associated with that position and chromsome
#'   \item LDL is the p_value of the trait
#'   \item gene is the gene associated with rsid
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"ldl"

#' TRIG GWAS summary dataset (made available from Lipid consortium data files)
#'
#' The dataset 'ldl' is a data.frame object with the summary results of a GWAS analysis.
#'
#' @docType data
#'
#' @usage data(trig)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item CHR is the chromosome number
#'   \item BP is the position of the chromosome
#'   \item rsid The SNP ID associated with that position and chromsome
#'   \item TRIGS is the p_value of the trait
#'   \item gene is the gene associated with rsid
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"trig"

#' TOTAL CHOLESTROL GWAS summary dataset (made available from Lipid consortium data files)
#'
#' The dataset 'tchol' is a data.frame object with the summary results of a GWAS analysis.
#'
#' @docType data
#'
#' @usage data(tchol)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item CHR is the chromosome number
#'   \item BP is the position of the chromosome
#'   \item rsid The SNP ID associated with that position and chromsome
#'   \item TOTAL_CHOLESTROL is the p_value of the trait
#'   \item gene is the gene associated with rsid
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"tchol"

#' LD Block data summarized in Approximately independent linkage disequilibrium blocks in human populations
#'
#' The dataset 'bppeur' is a data.frame object with the LD blocks for European population.
#'
#' @docType data
#'
#' @usage data(bppeur)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item chr is the chromosome number
#'   \item start is the start position of the chromosome
#'   \item stop is the end position of the chromosome
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"bppeur"

#' LD Block data summarized in Approximately independent linkage disequilibrium blocks in human populations
#'
#' The dataset 'bppasn' is a data.frame object with the LD blocks for asian population.
#'
#' @docType data
#'
#' @usage data(bppasn)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item chr is the chromosome number
#'   \item start is the start position of the chromosome
#'   \item stop is the end position of the chromosome
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"bppasn"

#' LD Block data summarized in Approximately independent linkage disequilibrium blocks in human populations
#'
#' The dataset 'bppafr' is a data.frame object with the LD blocks for European population.
#'
#' @docType data
#'
#' @usage data(bppafr)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item chr is the chromosome number
#'   \item start is the start position of the chromosome
#'   \item stop is the end position of the chromosome
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"bppafr"

#' Skeleton SNP's for ipheGWAS
#'
#' @docType data
#'
#' @usage data(skeltonsnps)
#'
#' @format A dataframe with Skeleton SNP's (simulated using plink 1.9)
#' \itemize{
#'   \item rsid is the SNP
#'   \item A1 is allele1
#'   \item A1 is allele2
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"skeltonsnps"


#' IBD GWAS summary dataset
#'
#' The dataset 'ibd' is a data.frame object with the summary results of a GWAS analysis.
#'
#' @docType data
#'
#' @usage data(ibd)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item CHR is the chromosome number
#'   \item BP is the position of the chromosome
#'   \item rsid The SNP ID associated with that position and chromsome
#'   \item A1
#'   \item A2
#'   \item beta is the p_value of the trait
#'   \item se
#'   \item P.value is the p_value of the trait
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"ibd"

#' UlcerativeColitis GWAS summary dataset
#'
#' The dataset 'UlcerativeColitis' is a data.frame object with the summary results of a GWAS analysis.
#'
#' @docType data
#'
#' @usage data(UlcerativeColitis)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item CHR is the chromosome number
#'   \item BP is the position of the chromosome
#'   \item rsid The SNP ID associated with that position and chromsome
#'   \item A1
#'   \item A2
#'   \item beta is the p_value of the trait
#'   \item se
#'   \item P.value is the p_value of the trait
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"UlcerativeColitis"

#' Wasisthipratio GWAS summary dataset
#'
#' The dataset 'Wasisthipratio' is a data.frame object with the summary results of a GWAS analysis.
#'
#' @docType data
#'
#' @usage data(Wasisthipratio)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item CHR is the chromosome number
#'   \item BP is the position of the chromosome
#'   \item rsid The SNP ID associated with that position and chromsome
#'   \item A1
#'   \item A2
#'   \item beta is the p_value of the trait
#'   \item se
#'   \item P.value is the p_value of the trait
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"Wasisthipratio"

#' bmi GWAS summary dataset
#'
#' The dataset 'bmi' is a data.frame object with the summary results of a GWAS analysis.
#'
#' @docType data
#'
#' @usage data(bmi)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item CHR is the chromosome number
#'   \item BP is the position of the chromosome
#'   \item rsid The SNP ID associated with that position and chromsome
#'   \item A1
#'   \item A2
#'   \item beta is the p_value of the trait
#'   \item se
#'   \item P.value is the p_value of the trait
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"bmi"

#' CrohnsDisease GWAS summary dataset
#'
#' The dataset 'CrohnsDisease' is a data.frame object with the summary results of a GWAS analysis.
#'
#' @docType data
#'
#' @usage data(CrohnsDisease)
#'
#' @format A dataframe with simulated GWAS results (simulated using plink 1.9)
#' \itemize{
#'   \item CHR is the chromosome number
#'   \item BP is the position of the chromosome
#'   \item rsid The SNP ID associated with that position and chromsome
#'   \item A1
#'   \item A2
#'   \item beta is the p_value of the trait
#'   \item se
#'   \item P.value is the p_value of the trait
#' }
#'
#' @keywords datasets
#' @return A data frame
#'
"CrohnsDisease"
